src/rpc/methods/eth.rs

```diff
@@ -948,6 +949,77 @@ pub fn new_eth_tx_from_signed_message<DB: Blockstore>(
     Ok(ApiEthTx { hash, ..tx })
 }
 
+/// Creates an Ethereum transaction from Filecoin message lookup. If `None` is passed for `tx_index`,
+/// it looks up the transaction index of the message in the tipset.
+/// Otherwise, it uses some index passed into the function.
+fn new_eth_tx_from_message_lookup<DB: Blockstore>(
+    ctx: &Ctx<DB>,
+    message_lookup: &MessageLookup,
+    tx_index: Option<u64>,
+) -> Result<ApiEthTx> {
+    let ts = ctx
+        .chain_store()
+        .load_required_tipset_or_heaviest(&message_lookup.tipset)?;
+
+    // This transaction is located in the parent tipset
+    let parent_ts = ctx
+        .chain_store()
+        .load_required_tipset_or_heaviest(ts.parents())?;
+
+    let parent_ts_cid = parent_ts.key().cid()?;
+
+    // Lookup the transaction index
+    let tx_index = {
+        if let Some(index) = tx_index {
+            index
+        } else {
+            let msgs = ctx.chain_store().messages_for_tipset(&parent_ts)?;
+            if let Some((i, _)) = msgs
+                .iter()
+                .enumerate()
+                .find(|(_, msg)| msg.cid() == message_lookup.message)
+            {
+                i as u64
+            } else {
+                bail!("cannot find the msg in the tipset")
+            }
+        }
+    };
+
+    let smsg = get_signed_message(ctx, message_lookup.message)?;
+
+    let state = StateTree::new_from_root(ctx.store().into(), ts.parent_state())?;
+
+    Ok(ApiEthTx {
+        block_hash: parent_ts_cid.into(),
+        block_number: (parent_ts.epoch() as u64).into(),
+        transaction_index: tx_index.into(),
+        ..new_eth_tx_from_signed_message(&smsg, &state, ctx.chain_config().eth_chain_id)?
+    })
+}
+
+fn get_signed_message<DB: Blockstore>(ctx: &Ctx<DB>, message_cid: Cid) -> Result<SignedMessage> {
+    let result: Result<SignedMessage, crate::chain::Error> =
+        crate::chain::message_from_cid(ctx.store(), &message_cid);
+
+    if let Ok(smsg) = result {
+        return Ok(smsg);
+    }
+
+    // We couldn't find the signed message, it might be a BLS message, so search for a regular message.
+    let result: Result<Message, crate::chain::Error> =
+        crate::chain::message_from_cid(ctx.store(), &message_cid);
+    match result {
+        Ok(msg) => Ok(SignedMessage::new_unchecked(
+            msg,
+            Signature::new_bls(vec![]),
+        )),
+        Err(err) => {
+            bail!("failed to find msg {}: {}", message_cid, err)
+        }
+    }
```

Comment:
```suggestion
       result.or_else(|_| {
        // We couldn't find the signed message, it might be a BLS message, so search for a regular message.
        let msg: Message = crate::chain::message_from_cid(ctx.store(), &message_cid)
            .with_context(|| format!("failed to find msg {}", message_cid))?;
        Ok(SignedMessage::new_unchecked(
            msg,
            Signature::new_bls(vec![]),
        ))
    })
```

Reply:
```
.with_context(|| format!("failed to find msg {message_cid}"))?;
```